File: 3686.number-of-stable-subsequences.cpp
#
# @lc app=leetcode id=3686 lang=cpp
#
# [3686] Number of Stable Subsequences
#

# @lc code=start
class Solution {
public:
    int countStableSubsequences(vector<int>& nums) {
        const int MOD = 1000000007;
        int n = nums.size();
        long long sum1[2] = {0, 0};
        long long sum2[2] = {0, 0};
        long long ans = 0;
        for(int i = 0; i < n; ++i) {
            int p = nums[i] % 2;
            int op = 1 - p;
            long long f1 = (1LL + sum1[op] + sum2[op]) % MOD;
            long long f2 = sum1[p] % MOD;
            ans = (ans + f1 + f2) % MOD;
            sum1[p] = (sum1[p] + f1) % MOD;
            sum2[p] = (sum2[p] + f2) % MOD;
        }
        return static_cast<int>(ans);
    }
};
# @lc code=end

File: 3686.number-of-stable-subsequences.py
#
# @lc app=leetcode id=3686 lang=python3
#
# [3686] Number of Stable Subsequences
#

# @lc code=start
class Solution:
    def countStableSubsequences(self, nums: List[int]) -> int:
        MOD = 10**9 + 7
        prev = [[0] * 2 for _ in range(2)]
        for num in nums:
            cur_p = num % 2
            curr = [[0] * 2 for _ in range(2)]
            for c in range(2):
                for p in range(2):
                    curr[c][p] = prev[c][p]
            curr[0][cur_p] = (curr[0][cur_p] + 1) % MOD
            sum_diff = (prev[0][1 - cur_p] + prev[1][1 - cur_p]) % MOD
            curr[0][cur_p] = (curr[0][cur_p] + sum_diff) % MOD
            sum_same = prev[0][cur_p]
            curr[1][cur_p] = (curr[1][cur_p] + sum_same) % MOD
            prev = curr
        total = 0
        for c in range(2):
            for p in range(2):
                total = (total + prev[c][p]) % MOD
        return total

# @lc code=end